oki doki vamo la

# Inserção e análise inicial dos dados

In [79]:
#Importação de bibliotecas
import pandas as pd
import numpy as np

Sobre o uso de generators: https://www.blog.pythonlibrary.org/2014/01/27/python-201-an-intro-to-generators/

In [118]:
#Importação dos datasets

df = pd.read_excel('IMF_Dataset_Modified.xlsx', dtype = {'Type_ER': 'bool'} )
pd.set_option('display.max_columns', None)
#pd.set_option('display.max_rows', None)

*Obs.: aqui eu estava fazendo alguns testes. Minha dúvida é: eu preciso mudar o dtype das colunas do meu dataframe ao importá-lo como .csv? Ou posso só transformar os valores em ```True``` e ```False```?*

In [119]:
a = {1: True, 2: False}
index = [0,1]

b = pd.DataFrame.from_dict(a, orient = 'index').reset_index()

b.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   index   2 non-null      int64
 1   0       2 non-null      bool 
dtypes: bool(1), int64(1)
memory usage: 146.0 bytes


*Um disclaimer sobre chunks*

In [120]:
#Pra caso eu vá tentar mexer com os chunks (já vi que na função pd.read_excel não dá pra mexer com esse trem de chunk)
#print(next(df))

In [121]:
#Quantidade de linhas e de colunas
df.shape

(3922, 95)

Separando em uma variável ```colunas_df``` os nomes das colunas do dataframe.

In [122]:
colunas_df = list(df.columns)
print(colunas_df)
print('\nQuantidade de colunas: ', len(colunas_df))

['Country #', 'Year', 'Country Name', 'Type_ER', 'Type_RR', 'Type_BBR', 'Type_DR', 'NSB_ER', 'NSB_RR', 'NSB_BBR', 'NSB_DR', 'YOI_ER', 'YOI_RR', 'YOI_BBR', 'YOI_DR', 'YOChange_ER', 'YOChange_RR', 'YOChange_BBR', 'YOChange_DR', 'rules_in_place_yes_no', 'quantity_of_rules_in_place', 'quantity_of_strictly_national', 'Key elements of fiscal rules', 'monit_nat_ER', 'monit_nat_RR', 'monit_nat_BBR', 'monit_nat_DR', 'monit_supnat_ER', 'monit_supnat_RR', 'monit_supnat_BBR', 'monit_supnat_DR', 'enforc_nat_ER', 'enforc_nat_RR', 'enforc_nat_BBR', 'enforc_nat_DR', 'enforc_sup_ER', 'enforc_sup_RR', 'enforc_sup_BBR', 'enforc_sup_DR', 'cov_nat_ER', 'cov_nat_RR', 'cov_nat_BBR', 'cov_nat_DR', 'cov_sup_ER', 'cov_sup_RR', 'cov_sup_BBR', 'cov_sup_DR', 'lb_nat_ER', 'lb_nat_RR', 'lb_nat_BBR', 'lb_nat_DR', 'lb_sup_ER', 'lb_sup_RR', 'lb_sup_BBR', 'lb_sup_DR', 'EC_nat_ER', 'EC_nat_RR', 'EC_nat_BBR', 'EC_nat_DR', 'EC_sup_ER', 'EC_sup_RR', 'EC_sup_BBR', 'EC_sup_DR', 'my_expc_agreggate', 'my_expc_ministries', 'my_e

In [123]:
df.head()

Country #  Year Country Name  Type_ER Type_RR Type_BBR Type_DR NSB_ER  \
0          1  1985      Andorra     True       -        -       -      -   
1          1  1986      Andorra     True       -        -       -      -   
2          1  1987      Andorra     True       -        -       -      -   
3          1  1988      Andorra     True       -        -       -      -   
4          1  1989      Andorra     True       -        -       -      -   

  NSB_RR NSB_BBR NSB_DR YOI_ER YOI_RR YOI_BBR YOI_DR YOChange_ER YOChange_RR  \
0      -       -      -      -      -       -      -           -           -   
1      -       -      -      -      -       -      -           -           -   
2      -       -      -      -      -       -      -           -           -   
3      -       -      -      -      -       -      -           -           -   
4      -       -      -      -      -       -      -           -           -   

  YOChange_BBR YOChange_DR rules_in_place_yes_no quantity_of_rules_in_place  \
0   2018, 2019  2018, 2019                    no                          0   
1   2018, 2019  2018, 2019                    no                          0   
2   2018, 2019  2018, 2019                    no                          0   
3   2018, 2019  2018, 2019                    no                          0   
4   2018, 2019  2018, 2019                    no                          0   

  quantity_of_strictly_national  \
0                             0   
1                             0   
2                             0   
3                             0   
4                             0   

                        Key elements of fiscal rules monit_nat_ER  \
0  Fiscal rules were implemented in 2014 and desi...            -   
1                                                NaN            -   
2                                                NaN            -   
3                                                NaN            -   
4                                                NaN            -   

  monit_nat_RR monit_nat_BBR monit_nat_DR monit_supnat_ER monit_supnat_RR  \
0            -             -            -               -               -   
1            -             -            -               -               -   
2            -             -            -               -               -   
3            -             -            -               -               -   
4            -             -            -               -               -   

  monit_supnat_BBR monit_supnat_DR enforc_nat_ER enforc_nat_RR enforc_nat_BBR  \
0                -               -             -             -              -   
1                -               -             -             -              -   
2                -               -             -             -              -   
3                -               -             -             -              -   
4                -               -             -             -              -   

  enforc_nat_DR enforc_sup_ER enforc_sup_RR enforc_sup_BBR enforc_sup_DR  \
0             -             -             -              -             -   
1             -             -             -              -             -   
2             -             -             -              -             -   
3             -             -             -              -             -   
4             -             -             -              -             -   

  cov_nat_ER cov_nat_RR cov_nat_BBR cov_nat_DR cov_sup_ER cov_sup_RR  \
0          -          -           -          -          -          -   
1          -          -           -          -          -          -   
2          -          -           -          -          -          -   
3          -          -           -          -          -          -   
4          -          -           -          -          -          -   

  cov_sup_BBR cov_sup_DR lb_nat_ER lb_nat_RR lb_nat_BBR lb_nat_DR lb_sup_ER  \
0           -          -         -         - 

In [26]:
df.tail()

Country #  Year Country Name Type_ER Type_RR Type_BBR Type_DR NSB_ER  \
3917        106  2017      Vietnam       1       1        1       1      0   
3918        106  2018      Vietnam       1       1        1       1      0   
3919        106  2019      Vietnam       1       1        1       1      0   
3920        106  2020      Vietnam       1       1        1       1      0   
3921        106  2021      Vietnam       1       1        1       1      0   

     NSB_RR NSB_BBR NSB_DR YOI_ER YOI_RR YOI_BBR YOI_DR YOChange_ER  \
3917      0       0      0   2016   2016    2016   2016  2016, 2021   
3918      0       0      0   2016   2016    2016   2016  2016, 2021   
3919      0       0      0   2016   2016    2016   2016  2016, 2021   
3920      0       0      0   2016   2016    2016   2016  2016, 2021   
3921      0       0      0   2016   2016    2016   2016  2016, 2021   

     YOChange_RR YOChange_BBR YOChange_DR rules_in_place_yes_no  \
3917  2016, 2021   2016, 2021  2016, 2021                   yes   
3918  2016, 2021   2016, 2021  2016, 2021                   yes   
3919  2016, 2021   2016, 2021  2016, 2021                   yes   
3920  2016, 2021   2016, 2021  2016, 2021                   yes   
3921  2016, 2021   2016, 2021  2016, 2021                   yes   

     quantity_of_rules_in_place quantity_of_strictly_national  \
3917                          4                             4   
3918                          4                             4   
3919                          4                             4   
3920                          4                             4   
3921                          4                             4   

     Key elements of fiscal rules monit_nat_ER monit_nat_RR monit_nat_BBR  \
3917                          NaN            1            1             1   
3918                          NaN            1            1             1   
3919                          NaN            1            1             1   
3920                          NaN            1            1             1   
3921                          NaN            1            1             1   

     monit_nat_DR monit_supnat_ER monit_supnat_RR monit_supnat_BBR  \
3917            1               -               -                -   
3918            1               -               -                -   
3919            1               -               -                -   
3920            1               -               -                -   
3921            1               -               -                -   

     monit_supnat_DR enforc_nat_ER enforc_nat_RR enforc_nat_BBR enforc_nat_DR  \
3917               -             1             1              1             1   
3918               -             1             1              1             1   
3919               -             1             1              1             1   
3920               -             1             1              1             1   
3921               -             1             1              1             1   

     enforc_sup_ER enforc_sup_RR enforc_sup_BBR enforc_sup_DR cov_nat_ER  \
3917             -             -              -             -          2   
3918             -             -              -             -          2   
3919             -             -              -             -          2   
3920             -             -              -             -          2   
3921             -             -              -             -          2   

     cov_nat_RR cov_nat_BBR cov_nat_DR cov_sup_ER cov_sup_RR cov_sup_BBR  \
3917          2           2          2          -          -           -   
3918          2           2          2          -          -           -   
3919          2           2          2          -          -           -   
3920          2           2          2          -          -           -   
3921          2           2          2          -          -           -   

     cov_sup_DR lb_nat_ER lb_na

## Tipos das colunas

A partir do método ```.info()``` nota-se que **várias colunas não foram importadas da melhor forma** para nós e são de **tipo ```object```.**

É necessário pensar qual deveria ser o tipo esperado de cada uma das colunas e verificar o que deve ser modificado para que todos os valores da coluna atendam ao critério do tipo esperado.

In [97]:
df.info(show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3922 entries, 0 to 3921
Data columns (total 95 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   Country #                      3922 non-null   int64 
 1   Year                           3922 non-null   int64 
 2   Country Name                   3922 non-null   object
 3   Type_ER                        3922 non-null   bool  
 4   Type_RR                        3922 non-null   object
 5   Type_BBR                       3922 non-null   object
 6   Type_DR                        3922 non-null   object
 7   NSB_ER                         3922 non-null   object
 8   NSB_RR                         3922 non-null   object
 9   NSB_BBR                        3922 non-null   object
 10  NSB_DR                         3922 non-null   object
 11  YOI_ER                         3922 non-null   object
 12  YOI_RR                         3910 non-null   object
 13  YOI

Dessa forma, o método ```describe()``` não é útil para fazer nenhuma análise.

In [91]:
df.describe()

Country #        Year     advanced     emerging  resource_rich  \
count  3922.000000  3922.00000  3922.000000  3922.000000    3922.000000   
mean     53.500000  2003.00000     0.301887     0.405660       0.219531   
std      30.602104    10.67844     0.459135     0.491082       0.413982   
min       1.000000  1985.00000     0.000000     0.000000       0.000000   
25%      27.000000  1994.00000     0.000000     0.000000       0.000000   
50%      53.500000  2003.00000     0.000000     0.000000       0.000000   
75%      80.000000  2012.00000     1.000000     1.000000       0.000000   
max     106.000000  2021.00000     1.000000     1.000000       1.000000   

                EU      federal  member_currency_union    region     
count  3922.000000  3922.000000            3922.000000  3922.000000  
mean      0.243498     0.132075               0.349312     2.784039  
std       0.429248     0.338616               0.476813     1.479787  
min       0.000000     0.000000               0.000000     0.000000  
25%       0.000000     0.000000               0.000000     2.000000  
50%       0.000000     0.000000               0.000000     2.000000  
75%       0.000000     0.000000               1.000000     4.000000  
max       1.000000     1.000000               1.000000     5.000000

Uma função útil para filtrar o dataset é ```df.select_dtypes[int64]```

O dicionário abaixo apresenta a **quantidade de valores ```null```** em cada coluna. (Ainda, muitas linhas do dataframe tem apenas o valor **```-```**)

In [22]:
dict(df.isnull().sum())

{'Country #': 0,
 'Year': 0,
 'Country Name': 0,
 'Type_ER': 0,
 'Type_RR': 0,
 'Type_BBR': 0,
 'Type_DR': 0,
 'NSB_ER': 0,
 'NSB_RR': 0,
 'NSB_BBR': 0,
 'NSB_DR': 0,
 'YOI_ER': 0,
 'YOI_RR': 12,
 'YOI_BBR': 0,
 'YOI_DR': 0,
 'YOChange_ER': 5,
 'YOChange_RR': 13,
 'YOChange_BBR': 2,
 'YOChange_DR': 2,
 'rules_in_place_yes_no': 0,
 'quantity_of_rules_in_place': 0,
 'quantity_of_strictly_national': 0,
 'Key elements of fiscal rules': 3816,
 'monit_nat_ER': 6,
 'monit_nat_RR': 6,
 'monit_nat_BBR': 7,
 'monit_nat_DR': 7,
 'monit_supnat_ER': 6,
 'monit_supnat_RR': 6,
 'monit_supnat_BBR': 6,
 'monit_supnat_DR': 6,
 'enforc_nat_ER': 9,
 'enforc_nat_RR': 6,
 'enforc_nat_BBR': 6,
 'enforc_nat_DR': 6,
 'enforc_sup_ER': 6,
 'enforc_sup_RR': 6,
 'enforc_sup_BBR': 6,
 'enforc_sup_DR': 6,
 'cov_nat_ER': 6,
 'cov_nat_RR': 6,
 'cov_nat_BBR': 6,
 'cov_nat_DR': 12,
 'cov_sup_ER': 0,
 'cov_sup_RR': 0,
 'cov_sup_BBR': 0,
 'cov_sup_DR': 0,
 'lb_nat_ER': 6,
 'lb_nat_RR': 6,
 'lb_nat_BBR': 6,
 'lb_nat_DR': 6

Obs.: note que, para a análise de algumas colunas, será necessário fazer uma adaptação dos dados (ex.: adaptar os valores para caberem em um índice de 0 a 1)

# Filtrando os dados


Há muitas colunas no dataframe que contém apenas valores '0' para Falso e '1' para verdadeiro. Outras, por sua vez, são números inteiros, outras floats.


## Alterando colunas para 'booleanas'

Para ser mais fácil a visualização e entendimento, irei transformar todas as colunas que, teoricamente, são binárias, em colunas booleanas com valores 'True' e 'False', somente.

Isso pode mudar a abordagem da elaboração dos gráficos mais à frente. **Link de referência para contagem de valores True/False: https://stackoverflow.com/questions/53415751/count-occurences-of-true-false-in-column-of-dataframe**.

Criei uma pequena planilha no excel com os nomes das colunas e uma segunda coluna informando se os valores esperados dela são binários ou não. Vou importar essa planilha 'binary_or_not'.

In [67]:
binary_or_not = pd.read_excel('binary_or_not.xlsx')
binary_or_not.head()

Country #  NB
0          Year  NB
1  Country Name  NB
2       Type_ER   B
3       Type_RR   B
4      Type_BBR   B

In [68]:
binary_or_not.shape

(94, 2)

Houve um problema com este dataframe que importamos. Ele tem 94 linhas, ao invés das 95 que deveria ter, porque a primeira linha do dataset se tornou os nomes das colunas (```Country #``` e ```NB```).



In [69]:
binary_or_not.loc[-1] = ['Country #', 'NB']  # adicionando os valores da primeira linha
binary_or_not = binary_or_not.sort_index()  # colocando em ordem pelo índice
binary_or_not.head()

Country #  NB
-1     Country #  NB
 0          Year  NB
 1  Country Name  NB
 2       Type_ER   B
 3       Type_RR   B

In [72]:
binary_or_not.index = binary_or_not.index + 1  # adicionando + 1 em todos os índices para corrigir os valores

In [75]:
#Renomeando as colunas usando um dicionário
binary_or_not.rename(columns = {'Country #': 'Name of column', 'NB':'Binary or not'}, inplace = True)

print(binary_or_not.head())

  Name of column Binary or not
0      Country #            NB
1           Year            NB
2   Country Name            NB
3        Type_ER             B
4        Type_RR             B


In [74]:
binary_or_not.shape

(95, 2)

**Dataset ```binary_or_not``` corrigido!**

Agora vou pegar uma **lista de colunas que são binárias e serão convertidas para colunas booleanas.**

In [165]:
df_colunas_booleanas = binary_or_not[binary_or_not['Binary or not'] == 'B']
colunas_booleanas = list(df_colunas_booleanas['Name of column'])
print(colunas_booleanas)
quant_cols_bool = len(colunas_booleanas)
print('\nHá ', quant_cols_bool, ' colunas booleanas.')

['Type_ER', 'Type_RR', 'Type_BBR', 'Type_DR', 'rules_in_place_yes_no', 'monit_nat_ER', 'monit_nat_RR', 'monit_nat_BBR', 'monit_nat_DR', 'monit_supnat_ER', 'monit_supnat_RR', 'monit_supnat_BBR', 'monit_supnat_DR', 'enforc_nat_ER', 'enforc_nat_RR', 'enforc_nat_BBR', 'enforc_nat_DR', 'enforc_sup_ER', 'enforc_sup_RR', 'enforc_sup_BBR', 'enforc_sup_DR', 'EC_nat_ER', 'EC_nat_RR', 'EC_nat_BBR', 'EC_nat_DR', 'EC_sup_ER', 'EC_sup_RR', 'EC_sup_BBR', 'EC_sup_DR', 'my_expc_agreggate', 'my_expc_ministries', 'my_expc_line-items', 'ind_body_sets_budg_assump', 'ind_body_monit_implement', 'fiscal_responsability_law', 'stab_over_cycle_nat', 'stab_over_cycle_sup', 'invest_excluded_nat', 'invest_excluded_sup', 'actv_EC_nat_ER', 'actv_EC_nat_RR', 'actv_EC_nat_BBR', 'actv_EC_nat_DR', 'actv_EC_sup_ER', 'actv_EC_sup_RR', 'actv_EC_sup_BBR', 'actv_EC_sup_DR', 'suspension_ER', 'suspension_RR', 'suspension_BBR', 'suspension_DR', 'advanced', 'emerging', 'low_income', 'resourse_rich', 'EU', 'federal', 'member_curre

### Teste:

Se eu mudar todos os valores de uma coluna para ```True``` ou ```False```, o tipo dela se torna bool?



In [148]:
from numpy import nan 
dataframe_teste = pd.DataFrame({ 'DeFatoBool': [True, False, True], 'NaoEhBoolVouMudar': [1, nan , 0]})
print(type(dataframe_teste))
dataframe_teste.head()


<class 'pandas.core.frame.DataFrame'>


DeFatoBool  NaoEhBoolVouMudar
0        True                1.0
1       False                NaN
2        True                0.0

In [149]:
dataframe_teste.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   DeFatoBool         3 non-null      bool   
 1   NaoEhBoolVouMudar  2 non-null      float64
dtypes: bool(1), float64(1)
memory usage: 155.0 bytes


In [150]:
dataframe_teste['NaoEhBoolVouMudar'] = dataframe_teste['NaoEhBoolVouMudar'].astype(bool)
dataframe_teste.head()


DeFatoBool  NaoEhBoolVouMudar
0        True               True
1       False               True
2        True              False

In [151]:
dataframe_teste.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   DeFatoBool         3 non-null      bool 
 1   NaoEhBoolVouMudar  3 non-null      bool 
dtypes: bool(2)
memory usage: 134.0 bytes


### Fim do teste

Conclusão: o ```.astype(bool)``` só muda para ```False``` o valor ```0``` especificamente. Qualquer outro valor, mesmo ```NaN``` e ```None```, são tratados como ```1```.

Como precisamos que todos os dados sejam apenas ```0``` ou ```1``` para passar para Bool, vou verificar quantos elementos ```-```, ```NaN```, ```1``` e ```0``` tem em cada coluna que é booleana (as colunas de interesse no momento). Lembrando que todas as colunas devem somar 3922. Se alguma coluna não somar 3922, tem algo errado.

In [185]:
# Criar uma função que conta a quantidade de diferentes valores em cada coluna de interesse

def count_keys(dataframe, coluna):
    vals_e_quants = {} # inicializando um dicionário vazio para guardar os diferentes valores e suas quants
    
    # Iterando pela coluna de interesse
    for row in dataframe[coluna]:
        
        # Se o valor atual já existir no dicionário, ele é incrementado em uma unidade
        if row in vals_e_quants.keys():
            vals_e_quants[row] += 1
        
        # Se o valor atual ainda não tiver sido contabilizado, ele passa a existir no dicionário com o valor 1
        else:
            vals_e_quants[row] = 1
    return vals_e_quants

type_rr = count_keys(df, 'Type_RR')
print(type_rr)

{'-': 3586, 1: 336}


In [186]:
quants = []
for coluna in colunas_booleanas:
    dic = count_keys(df, coluna)
    quants.append(dic)
print(quants)

KeyError: 'resourse_rich'

In [187]:
print([coluna for coluna in colunas_booleanas])

['Type_ER', 'Type_RR', 'Type_BBR', 'Type_DR', 'rules_in_place_yes_no', 'monit_nat_ER', 'monit_nat_RR', 'monit_nat_BBR', 'monit_nat_DR', 'monit_supnat_ER', 'monit_supnat_RR', 'monit_supnat_BBR', 'monit_supnat_DR', 'enforc_nat_ER', 'enforc_nat_RR', 'enforc_nat_BBR', 'enforc_nat_DR', 'enforc_sup_ER', 'enforc_sup_RR', 'enforc_sup_BBR', 'enforc_sup_DR', 'EC_nat_ER', 'EC_nat_RR', 'EC_nat_BBR', 'EC_nat_DR', 'EC_sup_ER', 'EC_sup_RR', 'EC_sup_BBR', 'EC_sup_DR', 'my_expc_agreggate', 'my_expc_ministries', 'my_expc_line-items', 'ind_body_sets_budg_assump', 'ind_body_monit_implement', 'fiscal_responsability_law', 'stab_over_cycle_nat', 'stab_over_cycle_sup', 'invest_excluded_nat', 'invest_excluded_sup', 'actv_EC_nat_ER', 'actv_EC_nat_RR', 'actv_EC_nat_BBR', 'actv_EC_nat_DR', 'actv_EC_sup_ER', 'actv_EC_sup_RR', 'actv_EC_sup_BBR', 'actv_EC_sup_DR', 'suspension_ER', 'suspension_RR', 'suspension_BBR', 'suspension_DR', 'advanced', 'emerging', 'low_income', 'resourse_rich', 'EU', 'federal', 'member_curre

# Gráficos: dataset do FMI

Aprender:
- como exportar imagens do jupyter notebook para o computador
- ferramentas para alteração do design dos gráficos


Defina funções. Use try e except

In [9]:
#Imagem com três gráficos de linha, um para cada tipo de país (sub-desenvolvido, emergente, desenvolvido), 
#indicando a quantidade de países com algum tipo de regra fiscal ao longo do tempo (1995 - 2021)

In [10]:
#Imagem com dois gráficos de linha indicando a quantidade de países que contém pelo menos um tipo 
#de regra fiscal nacional e supranacional ao longo do tempo (1995 - 2021)

In [11]:
#Imagem com quatro grupos de três gráficos de barra, um para cada tipo de país (sub-desenvolvido, emergente, desenvolvido), 
#indicando a quantidade de países com cada tipo de regra fiscal (eixo X: ER, RR, BBR, DR; eixo Y: quantidade, legenda: 
#sub-desenvolvido, emergente, desenvolvido)

Produção de uma imagem com FISCAL RULE FEATURES: 6 grupos de 3 gráficos de linha ao longo do tempo (separando por países sub-desenvolvidos, emergentes e avançados) 
1. Porcentagem de países do mesmo grupo de renda com independent body monitoring the budget
1. Porcentagem de países do mesmo grupo de renda com fiscal responsability law
1. Porcentagem de países do mesmo grupo de renda com cyclical adjustment of the balance
1. Porcentagem de países do mesmo grupo de renda com investment spending excluded from the target
1. Coverage dos países de um mesmo grupo de renda (valor ajustado para um índice de 0 a 1, já que pode assumir valores de 0 a 2,5)
1. Enforcement dos países de um mesmo grupo de renda (valor ajustado para um índice de 0 a 1, a partir da soma das colunas de monitoring of compliance outside government (0 ou 1) e formal enforcement procedures (0 ou 1) 

Outra ideia: fazer a mesma coisa mas para apenas UM ANO e com 6 grupos de 3 gráficos de barra (pode ser 2021, 2008, etc, se houverem dados).


In [12]:
#Cálculos: 1. Porcentagem de países do mesmo grupo de renda com independent body monitoring the budget

In [13]:
#Cálculos: 2. Porcentagem de países do mesmo grupo de renda com fiscal responsability law

In [14]:
#Cálculos: 3. Porcentagem de países do mesmo grupo de renda com cyclical adjustment of the balance

In [15]:
#Cálculos: 4. Porcentagem de países do mesmo grupo de renda com investment spending excluded from the target

In [16]:
#Cálculos: 5. Coverage dos países de um mesmo grupo de renda (valor ajustado para um índice de 0 a 1, já que pode 
#assumir valores de 0 a 2,5)

In [17]:
#Cálculos: 6. Enforcement dos países de um mesmo grupo de renda (valor ajustado para um índice de 0 a 1, a partir da soma das colunas de monitoring of compliance outside government (0 ou 1) e 
#formal enforcement procedures (0 ou 1) 

# Gráficos: dataset do Fiscal Rules Strength Index

Aprender a calcular os índices de força
Referência: pasta "Pesquisas semelhantes à minha"